### PR TITLE
Mseed: spread time shift out through individual data blockettes

### PR DESCRIPTION
**Pull request by ghfbsd**
### What does this PR do?

This adds a new routine to utils.io.mseed called spread_time_over_file().  I wrote the routine to fix timing problems with mseed data generated during an old field experiment that had clock drift that caused day-long stretches of continuous data to accumulate an apparent time gap of about 0.5 sample interval by the end of the file.  By taking that gap interval and spreading it progressively over all the blockettes in the file, the continuity of the data can then be guaranteed.

The mseed data itself is not changed; only the headers.  The header modification follows the methods used in utils.io.mseed.shift_time_of_file().

Wrote unit test to test basic functionality.  Also successfully processed 11 Gb of historic data and removed the apparent gaps in it.  

### Why was it initiated?  Any relevant Issues?

Needed a new feature to fix existing data (see description above).

### PR Checklist
/ = not applicable
- [x] Correct base branch selected? `master` for new features, `maintenance_...` for bug fixes
- [x] This PR is not directly related to an existing issue (which has no PR yet).
- [x] While the PR is still work-in-progress, the `no_ci` label can be added to skip CI builds
- [x] If the PR is making changes to documentation, docs pages can be built automatically.
      Just add the `build_docs` tag to this PR.
      Docs will be served at [docs.obspy.org/pr/{branch_name}](https://docs.obspy.org/pr/) (do not use master branch).
      Please post a link to the relevant piece of documentation.
- [/] If all tests including network modules (e.g. `clients.fdsn`) should be tested for the PR,
      just add the `test_network` tag to this PR.
- [x] All tests still pass.
- [x] Any new features or fixed regressions are covered via new tests.
- [x] Any new or changed features are fully documented.
- [x] Significant changes have been added to `CHANGELOG.txt` .
- [x] First time contributors have added your name to `CONTRIBUTORS.txt` .
- [/] If the changes affect any plotting functions you have checked that the plots
      from all the CI builds look correct. Add the "upload_plots" tag so that plotting 
      outputs are attached as artifacts. 
- [/] New modules, add the module to `CODEOWNERS` with your github handle
- [x] Add the `ready for review` tag when you are ready for the PR to be reviewed.

### CODE REVIEW

**Comment by ghfbsd** [comment]
Something is broken in the test action.  There are over 30,000 errors, all of which relate to the version sqlalchemy check, one of which reads:

FAILED signal/tests/test_trigger.py::TestTrigger::test_coincidence_trigger_with_similarity_checking - pkg_resources.VersionConflict: (SQLAlchemy 2.0.4 (/usr/share/miniconda3/envs/test/lib/python3.8/site-packages), Requirement.parse('sqlalchemy<2'))

This has nothing to do with changes that I made, so proceeding is impossible.

**Comment by d-chambers** [comment]
Hi @ghfbsd,

> This has nothing to do with changes that I made, so proceeding is impossible.

Yes, definitely something with the wrong specified or installed version. I can probably find some time to take a look this weekend or early next week.

**Comment by filefolder** [comment]
> Hi @ghfbsd,
> 
> > This has nothing to do with changes that I made, so proceeding is impossible.
> 
> Yes, definitely something with the wrong specified or installed version. I can probably find some time to take a look this weekend or early next week.

i think this https://github.com/obspy/obspy/pull/3269 fixes it, if you want to have a quick review. it seems that the current stop-gap solution of suppressing the warnings isn't working for any PRs

**Comment by ghfbsd** [comment]
When you get this sorted, please let me know, and then advise on how to re-do the test action for the pull request.  Otherwise I guess I'll have to abandon it and do another one (3rd time may be a charm).

**Comment by d-chambers** [comment]
Ok @ghfbsd, you should be able to rebase this branch to master then force push (no need to open a new PR).

**Comment by megies** [comment]
Curious about the use case for this, but I guess no reason not to add this functionality, I'll leave some comments in the code changes

**Comment by ghfbsd** [comment]
> Curious about the use case for this, but I guess no reason not to add this functionality, I'll leave some comments in the code changes

The background.   This data was recorded on 1990s era RefTek 72A-02 dataloggers, having less stable clocks than modern hardware.  The data was then processed into day-long PASSCAL SEGY format for QC, and then to mseed for archiving.  When the dataset was excavated in 2023 only the day-long files, in mseed format, were available.  After 24 hours, the times in the last mseed blockette in one file, and the first mseed blockette in the next file had a 0.5 sample gap or overlap, accumulated due to clock drift, and frozen into the timing data by the SEGY format conversion.  Hence the need to spread the clock drift across each mseed blockette in the 24 hour file to eliminate the apparent gap/overlaps at the end of a day.

File: obspy/io/mseed/tests/test_mseed_util.py [hunk 1]
```diff
@@ -648,6 +648,26 @@ def test_time_shifting(self, testdata):
             st_before[0].stats.starttime -= 33.3
             assert st_before == st_after
 
+            # Test spread_time_over_file().
```
Comment (megies) [inline obspy/io/mseed/tests/test_mseed_util.py:651]: In general it's better to have tests for separate functionality also in separate test cases, so please put a test method around this part.

File: CHANGELOG.txt [hunk 2]
```diff
@@ -10,6 +10,9 @@ Changes:
    * set origin evaluation status to "rejected" if nonlinloc reports the
      location run as "ABORTED", "IGNORED" or "REJECTED" (see #3230)
    * store "NLLOC" info header line in event and origin comments (see #3230)
+ - obspy.io.mseed.spread_time_over_file:
+   * new routine to spread a time interval progressively across all mseed
+     blockettes in a file
```
Comment (megies) [inline CHANGELOG.txt:15]: please add PR number in brackets

Comment (megies) [reply CHANGELOG.txt:15]: oh and please keep alphabetic order (put before nlloc change)

File: obspy/io/mseed/tests/test_mseed_util.py [hunk 3]
```diff
@@ -648,6 +648,26 @@ def test_time_shifting(self, testdata):
...
+            filename = testdata['BW.BGLD.__.EHE.D.2008.001.first_10_records']
+            # Shift by 0.9 seconds (0.1 s gap between each blockette in file)
+            util.spread_time_over_file(filename, output_filename, 9000)
+            st_before = _read_mseed(filename)
+            st_after = _read_mseed(output_filename)
+            st_before_et = st_before[0].stats.endtime + 0.9
+            st_after_et = st_after[len(st_after)-1].stats.endtime
```
Comment (megies) [inline obspy/io/mseed/tests/test_mseed_util.py:658]: `st_after[-1]` for simplicity

**Comment by megies** [comment]
I pushed some minor wording changes @ghfbsd, please pull before doing further adjustments (or do `git pull --rebase` if you already have more local commits)

**Comment by megies** [comment]
You could also add this function here: https://github.com/obspy/obspy/blob/ebdcda16a9ee0f99b7512cd983eab3d11b8f6393/obspy/io/mseed/__init__.py#L126

Which renders here:
https://docs.obspy.org/master/packages/obspy.io.mseed.html

**Review by megies — APPROVED** [review]
Some minor comments, otherwise looks good to me

**Comment by megies** [comment]
Generated docs for this can be reviewed here:
https://docs.obspy.org/pr/mseed-newtool/packages/autogen/obspy.io.mseed.util.spread_time_over_file.html

**Comment by ghfbsd** [comment]
Requested changes made and all local tests passed.

Had to remove the ready_for_review tag to surmount build error declared by setuptools, to wit:

```
    File "/private/var/folders/9x/4qc212848xs7zc0059_zk4hr0000gr/T/pip-build-env-rzrlvsce/overlay/lib/python3.8/site-packages/setuptools/_vendor/packaging/version.py", line 197, in __init__
      raise InvalidVersion(f"Invalid version: '{version}'")
  setuptools.extern.packaging.version.InvalidVersion: Invalid version: 'ready-for-review.post0-3.g8fb533d31b.dirty.ghfbsd.mseed.newtool'
```

Now a conflict in CHANGELOG.txt due to request that I reorder the entries as PR comments requested.

Let me know how to proceed.

**Comment by megies** [comment]
I rebased and force pushed, will merge now. Thanks @ghfbsd :rocket:

**Comment by megies** [comment]
> Had to remove the ready_for_review tag to surmount build error declared by setuptools, to wit:
> 
> ```
>     File "/private/var/folders/9x/4qc212848xs7zc0059_zk4hr0000gr/T/pip-build-env-rzrlvsce/overlay/lib/python3.8/site-packages/setuptools/_vendor/packaging/version.py", line 197, in __init__
>       raise InvalidVersion(f"Invalid version: '{version}'")
>   setuptools.extern.packaging.version.InvalidVersion: Invalid version: 'ready-for-review.post0-3.g8fb533d31b.dirty.ghfbsd.mseed.newtool'
> ```

This is really weird.. :sweat_smile: 
Hmm, I see you have a tag set on your fork with name "ready-for-review" that mustve been by mistake and our version number generation picks that up for the main version number and that's obviously no good (should be "1.4.0" from the last release tag)

**Pull request merged**